src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -1861,4 +1866,53 @@ void InterpreterCreateQuery::addColumnsDescriptionToCreateQueryIfNecessary(ASTCr
     }
 }
 
+void InterpreterCreateQuery::processSQLSecurityOption(ContextPtr context_, ASTSQLSecurity & sql_security, bool is_attach)
+{
+    if (!sql_security.type.has_value())
+    {
+        /// SQL SECURITY/DEFINER options wasn't specified, will use defaults.
+        String default_security = context_->getSettingsRef().default_view_sql_security;
+
+        if (default_security == "NONE")
+            sql_security.type = ASTSQLSecurity::Type::NONE;
+        else if (default_security == "INVOKER")
+            sql_security.type = ASTSQLSecurity::Type::INVOKER;
+        else if (default_security == "DEFINER")
+        {
+            sql_security.type = ASTSQLSecurity::Type::DEFINER;
+            String default_definer = context_->getSettingsRef().default_view_definer;
+            if (default_definer == "CURRENT_USER")
+                sql_security.is_definer_current_user = true;
+            else
+                sql_security.definer = std::make_shared<ASTUserNameWithHost>(default_definer);
+        }
+        else
+            throw Exception(ErrorCodes::SYNTAX_ERROR, "Expected default_view_sql_security = <DEFINER | INVOKER | NONE>. Got {}", default_security);
+    }
+
+    const auto current_user_name = context_->getUserName();
+    if (sql_security.is_definer_current_user)
+    {
+        if (current_user_name.empty())
+            /// This can happen only when attaching a view for the first time after migration and with `CURRENT_USER` default.
+            sql_security.type = ASTSQLSecurity::Type::NONE;
+        else if (sql_security.definer)
+            sql_security.definer->replace(current_user_name);
+        else
+            sql_security.definer = std::make_shared<ASTUserNameWithHost>(current_user_name);
+    }
+    else if (sql_security.definer && !is_attach)
+    {
+        const auto definer_name = sql_security.definer->toString();
+
+        /// Validate that definer exists.
+        context_->getAccessControl().getID<User>(definer_name);
+        if (definer_name != current_user_name)
+            context_->checkAccess(AccessType::SET_DEFINER, definer_name);
+    }
+
+    if (sql_security.type == ASTSQLSecurity::Type::NONE && !is_attach)
+        context_->checkAccess(AccessType::ALLOW_SQL_SECURITY_NONE);
+}
```

Comment:
This function is quite hard to read because the conditions are intertwined. Maybe it's better to split these conditions into several more or less independent parts:

```
/// Apply defaults if it's an ATTACH command
if (is_attach)
{
    ...
    return;
}

/// Apply default from default_view_sql_security.
if (!sql_security.type)
    ... (set sql_security.type)

/// Apply default from default_view_definer.
if (sql_security.type == ASTSQLSecurity::Type::DEFINER)
    ... (set sql_security.definer & sql_security.is_definer_current_user)

/// If "DEFINER CURRENT_USER" is used we need to replace "CURRENT USER" with the name of the current user.
if (sql_security.is_definer_current_user)
    ...

/// Check access
...
```

Reply:
Your code snippet describes a different behavior. I've added some more comments, so hopefully, it will be less confusing.